### ISSUE TITLE
pylint.extensions.docparams's weird behavior with multiple-words argument type descriptions using Numpy style

### ISSUE THREAD

**Issue by franciscoadasme** [PyCQA/pylint#1383]
### Steps to reproduce
1. Write a function with two arguments.
2. Write a docstring using the Numpy convention marking the second argument as optional.

#### Example

```python
def do_something(param, param2='all'):
    """Do something.

    Parameters
    ----------
    param : str
        Description.
    param2 : str, optional
        Description (the default is 'all').

    Returns
    -------
    int
        Description.

    """
    print param, param2
    return 1
```

Note that this is supposed to be a valid docstring in the Numpy style.

### Current behavior
```
> pylint test.py
************* Module test
W:  4, 0: "param2" missing or differing in parameter documentation (missing-param-doc)
W:  4, 0: "param2" missing or differing in parameter type documentation (missing-type-doc)
```

If I remove the optional flag, the warnings disappear. The warning also happens when using anything other than a single word as parameter type (e.g., list of str) for `param2`. However, if I change `param` type description to optional or anything besides one word, the warning disappears again.

This problem does not occur when there is one argument. But if I add one or more arguments, different combinations of the above either succeed or don't.

### Expected behavior
No warning at all.

### pylint --version output
```
pylint 1.6.5,
astroid 1.4.9
Python 2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 20:53:40) [MSC v.1500 64 bit (AMD64)]
```
I'm on Windows 10 by the way.

**Comment by AWhetter** [comment]
Not sure when I'll have time to test this but hopefully the solution is simply to add `(?:, optional)?` after the optional type declaration of `NumpyDocstring.re_param_line`.
As for the problem with using multiple words in a parameter type, Sphinx 1.5 added syntax for container types and multiple types (http://www.sphinx-doc.org/en/stable/domains.html#info-field-lists) and this is what docparams looks for. So a list of strings can be documented with `list(str)` or `list[str]`.

**Comment by franciscoadasme** [comment]
@AWhetter Thank you for the quick reply. Hopefully you will be able to fix it soon.

Regarding the parameter type, either `list(str)` or `list[str]` indeed work, but I don't work whether it's appropriate within the Numpy style.

**Comment by franciscoadasme** [comment]
I forgot to tell you that documenting multiple types for a parameter/return also fails:

```python
def do_something(param='all', param2=1):
    """Do something.

    Parameters
    ----------
    param : list(str)
        Description.
    param2 : int or str
        Description.

    Returns
    -------
    float or str
        Description.

    """
    print param
    print param2
    return 1.0
```

pylint produces warnings W9003, W9004 and W9007 for this code regarding `param2` and the returning type. I think `or` is the suitable delimiter, or there is another way?

**Comment by AWhetter** [comment]
The multiple types issue has already been fixed! (#1137)

**Comment by dpizetta** [comment]
Hi can you extent this solution to GoogleStyle? We have the same problem.

**Docstring:**

- "include_parent (bool, optional): Defaults to False. Include parent object"

**Warning:**

- include_parent" missing in parameter documentation (missing-param-doc)

Tks
